src/bucket/BucketManager.h

```diff
@@ -114,11 +142,32 @@ class BucketManager : NonMovableOrCopyable
     // worker threads. Very carefully.
     virtual std::shared_ptr<Bucket>
     adoptFileAsBucket(std::string const& filename, uint256 const& hash,
-                      size_t nObjects, size_t nBytes) = 0;
+                      size_t nObjects, size_t nBytes,
+                      MergeKey* mergeKey = nullptr) = 0;
 
     // Return a bucket by hash if we have it, else return nullptr.
     virtual std::shared_ptr<Bucket> getBucketByHash(uint256 const& hash) = 0;
 
+    // Get a reference to a merge-future that's either running (or finished
+    // somewhat recently) from either a map of the std::shared_futures doing the
+    // merges and/or a set of records mapping merge inputs to outputs and the
+    // set of outputs held in the BucketManager. Returns an invalid future if no
+    // such future can be found or synthesized.
+    virtual std::shared_future<std::shared_ptr<Bucket>>
+    getMergeFuture(MergeKey const& key) = 0;
+
+    // Add a reference to a merge _in progress_ (not yet adopted as a file) to
+    // the BucketManager's internal map of std::shared_futures doing merges.
+    // There is no corresponding entry-removal API: the std::shared_future will
+    // be removed from the map when the merge completes and the output file is
+    // adopted.
+    virtual void
+    putMergeFuture(MergeKey const& key,
+                   std::shared_future<std::shared_ptr<Bucket>>) = 0;
+
+    // Drop all references to merge futures in progress.
+    virtual void clearMergeFutures() = 0;
```

Comment:
should that be called **ForTesting? It seems a bit dangerous

Reply:
Yeah fair. I'll debug-guard it too.